modules/sitemaps/sitemaps.php

```diff
@@ -1,792 +1,470 @@
-	$cur_datetime = current_time( 'mysql', true );
-
-	$query = $wpdb->prepare( "
-		SELECT p.ID, p.post_title, p.post_type, p.post_date, p.post_name, p.post_date_gmt, GROUP_CONCAT(t.name SEPARATOR ', ') AS keywords
-		FROM
-			$wpdb->posts AS p LEFT JOIN $wpdb->term_relationships AS r ON p.ID = r.object_id
-			LEFT JOIN $wpdb->term_taxonomy AS tt ON r.term_taxonomy_id = tt.term_taxonomy_id AND tt.taxonomy = 'post_tag'
-			LEFT JOIN $wpdb->terms AS t ON tt.term_id = t.term_id
-		WHERE
-			post_status='publish' AND post_type IN ( {$post_types_in_string} ) AND post_date_gmt > (%s - INTERVAL 2 DAY)
-		GROUP BY p.ID
-		ORDER BY p.post_date_gmt DESC LIMIT %d", $cur_datetime, $limit );
-
-	// URL to XSLT
-	$xsl = get_option( 'permalink_structure' ) ? home_url( 'news-sitemap.xsl' ) : home_url( '/?jetpack-news-sitemap-xsl=true' );
-
-	// Unless it's zh-cn for Simplified Chinese or zh-tw for Traditional Chinese,
-	// trim national variety so an ISO 639 language code as required by Google.
-	$language_code = strtolower( get_locale() );
-	if ( in_array( $language_code, array( 'zh_tw', 'zh_cn' ) ) ) {
-		$language_code = str_replace( '_', '-', $language_code );
-	} else {
-		$language_code = preg_replace( '/(_.*)$/i', '', $language_code );
-	}
+		if ( isset( $request['sitemap_uri'] ) ) {
 
-	header( 'Content-Type: application/xml' );
-	ob_start();
-	echo '<?xml version="1.0" encoding="UTF-8"?>' . "\n";
-	echo '<?xml-stylesheet type="text/xsl" href="' . esc_url( $xsl ) . '"?>' . "\n";
-	echo '<!-- generator="jetpack-' . JETPACK__VERSION . '" -->' . "\n";
-	?>
-	<!-- generator="jetpack" -->
-	<urlset xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
-	        xsi:schemaLocation="http://www.sitemaps.org/schemas/sitemap/0.9 http://www.sitemaps.org/schemas/sitemap/0.9/sitemap.xsd"
-	        xmlns="http://www.sitemaps.org/schemas/sitemap/0.9"
-	        xmlns:news="http://www.google.com/schemas/sitemap-news/0.9"
-	        xmlns:image="http://www.google.com/schemas/sitemap-image/1.1"
-		>
-		<?php
-		$posts = $wpdb->get_results( $query );
-		foreach ( $posts as $post ):
-			setup_postdata( $post );
+			if ( defined( 'WP_DEBUG' ) && ( true === WP_DEBUG ) ) {
+				$this->logger->report( 'sitemap uri:  "' . $request['sitemap_uri'] . '"' );
+				$this->logger->report( 'sitemap name: "' . $request['sitemap_name'] . '"' );
+			}
 
 			/**
-			 * Filter condition to allow skipping specific posts in news sitemap.
+			 * Filter the content type used to serve the sitemap XML files.
 			 *
 			 * @module sitemaps
 			 *
 			 * @since 3.9.0
 			 *
-			 * @param bool $skip Current boolean. False by default, so no post is skipped.
-			 * @param WP_POST $post Current post object.
+			 * @param string $xml_content_type By default, it's 'text/xml'.
 			 */
-			if ( apply_filters( 'jetpack_sitemap_news_skip_post', false, $post ) ) {
-				continue;
+			$xml_content_type = apply_filters( 'jetpack_sitemap_content_type', 'text/xml' );
```

Comment:
That's a good question.

The sitemap protocol does not specify what MIME types should be used, and I couldn't find any indication of what crawlers prefer (I assume they don't really care). So I fell back on [RFC 3023](https://www.ietf.org/rfc/rfc3023.txt) that says... a lot of things, but basically for documents ``text/xml'' is preferred if the source is human readable and ``application/xml'' if not. Anyway by that reasoning I thought the sitemaps themselves could be ``text/xml'', since they are pretty readable as is, while the stylesheets should be ``application/xml''. But looking at the code now I got the two mixed up. :) I fixed that- now stylesheets use ``application'' and all the sitemaps use the filter value which defaults to ``text''.

I'm not really convinced that this should be configurable with a filter, though, and I only left that in for backward compatibility. There are only two valid choices for the MIME type.